File: topi/python/topi/nn/tunable_sparse.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=invalid-name, unused-variable, too-many-locals
# pylint: disable=unused-argument, redefined-builtin
"""Tunable sparse operators"""
from __future__ import absolute_import as _abs
from collections import namedtuple
import tvm

from .pad import pad
from .util import get_pad_tuple

@tvm.target.generic_func
def sparse_conv2d(input, filter_data, filter_indices, filter_indptr, kernel_size, strides, padding, dilation, layout='NCHW', out_dtype=None):
    # search platform specific declaration first
    # default declaration
    if layout == 'NCHW':
        return sparse_conv2d_nchw(input, filter_data, filter_indices, filter_indptr, kernel_size, strides, padding, dilation, out_dtype)
    raise ValueError("not support this layout {} yet".format(layout))

def sparse_conv2d_nchw(Input, Filter_data, Filter_indices, Filter_indptr, kernel_size, stride, padding, dilation, out_dtype=None):
    if out_dtype is None:
        out_dtype = Input.dtype
    assert isinstance(kernel_size, int) or len(kernel_size) == 2
    assert isinstance(stride, int) or len(stride) == 2
    assert isinstance(dilation, int) or len(dilation) == 2
    if isinstance(kernel_size, int):
        kernel_h = kernel_w = kernel_size
    else:
        kernel_h, kernel_w = kernel_size

    if isinstance(stride, int):
        stride_h = stride_w = stride
    else:
        stride_h, stride_w = stride

    if isinstance(dilation, int):
        dilation_h = dilation_w = dilation
    else:
        dilation_h, dilation_w = dilation

    batch, in_channel, in_height, in_width = Input.shape
    if len(Filter_data.shape) == 1:
        num_filter = Filter_indptr.shape[0] - 1
    elif len(Filter_data.shape) == 3:
        num_filter = (Filter_indptr.shape[0] - 1) * Filter_data.shape[1]
    else:
        raise ValueError("Only CSR or BSR supported")
    channel = in_channel
    # compute the output shape
    dilated_kernel_h = (kernel_h - 1) * dilation_h + 1
    dilated_kernel_w = (kernel_w - 1) * dilation_w + 1
    pad_top, pad_left, pad_down, pad_right = get_pad_tuple(
        padding, (dilated_kernel_h, dilated_kernel_w))
    out_channel = num_filter
    out_height = simplify((in_height - dilated_kernel_h + pad_top + pad_down) // stride_h + 1)
    out_width = simplify((in_width - dilated_kernel_w + pad_left + pad_right) // stride_w + 1)
    # compute graph
    pad_before = [0, 0, pad_top, pad_left]
    pad_after = [0, 0, pad_down, pad_right]
    if pad_top != 0 or pad_left != 0 or pad_down != 0 or pad_right != 0:
        temp = pad(Input, pad_before, pad_after, name="pad_temp")
    else:
        temp = Input

    """
    '''
    im2col_out_shape = (batch, in_channel, kernel_h, kernel_w, out_height, out_width)
    def f(batch, c, h, w , out_h, out_w):
        return temp[batch, c, h + stride_h*out_h, w + stride_w*out_w]
    im2col_input = tvm.compute(im2col_out_shape, f, name='col')
    im2col_input = topi.transpose(im2col_input, axes=[0, 4, 5, 1, 2, 3])
    im2col_input = topi.reshape(im2col_input, newshape = (batch*out_height*out_width, in_channel*kernel_h*kernel_w))
    '''
    im2col_input = tvm.compute([batch * out_height * out_width, in_channel * kernel_h * kernel_w], lambda i, j:
            temp[i // (out_height*out_width)][j // (kernel_h*kernel_w)][i // out_width % out_height*stride_h + j // kernel_w % kernel_h][i % out_width*stride_w + j % kernel_w], name='col')
    '''
    im2col_input = tvm.compute((batch, out_height, out_width, in_channel, kernel_h, kernel_w), lambda n, oh, ow, ci, hk, wk: \
            temp[n][ci][oh*stride_h+hk][ow*stride_w+wk], name='col')
    im2col_input = topi.reshape(im2col_input, newshape=(batch*out_height*out_width, in_channel*kernel_h*kernel_w))
    '''

    output = topi.nn.sparse_dense(im2col_input, Filter_data, Filter_indices, Filter_indptr)
    output = topi.transpose(output, axes=[1, 0])
    output_shape = (batch, out_channel, out_height, out_width)
    output = topi.reshape(output, newshape=output_shape)

    '''
    s = tvm.create_schedule(output.op)
    print(tvm.lower(s, [Input, Filter_data, Filter_indices, Filter_indptr, output], simple_mode=True))
    '''
    """

    im2col_out_shape = (batch, in_channel, kernel_h, kernel_w, out_height, out_width)
    def f(b, c, h, w , out_h, out_w):
        return temp[b, c, h + stride_h*out_h, w + stride_w*out_w]
    im2col_input = tvm.compute(im2col_out_shape, f, name='col')
    im2col_input = topi.reshape(im2col_input, newshape=(batch*in_channel*kernel_h*kernel_w, out_height*out_width))

    output = _sparse_dense_bsrmm_v2(im2col_input, Filter_data, Filter_indices, Filter_indptr)
    output_shape = (batch, out_channel, out_height, out_width)
    output = topi.reshape(output, newshape=output_shape)

    return output

File: src/relay/op/nn/tunable_sparse.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file tunable_sparse.cc
 * \brief Tunable sparse operators
 */
#include <tvm/data_layout.h>
#include <tvm/ir_pass.h>
#include <tvm/relay/op.h>
#include <tvm/relay/attrs/nn.h>
#include <vector>

#include "../../pass/alter_op_layout.h"
#include "tunable_sparse.h"

namespace tvm {
namespace relay {

template<typename T>
Array<Array<Layout > > SparseConvInferCorrectLayout(
    const Attrs& attrs,
    const Array<Layout>& new_in_layouts,
    const Array<Layout>& old_in_layouts,
    const Array<Array<IndexExpr>> &old_in_shapes) {
  const T* params = attrs.as<T>();

  return Array<Array<Layout> >{{params->data_layout, "BRC", "B", "P"},
                               {params->out_layout == "" ?
                                    params->data_layout : params->out_layout}};
}

// relay.nn.sparse_conv2d
TVM_REGISTER_NODE_TYPE(SparseConv2DAttrs);


// Positional relay function to create conv2d operator
// used by frontend FFI.
Expr MakeSparseConv2D(Expr data,
                Expr weight_data,
                Expr weight_indices,
                Expr weight_indptr,
                Array<IndexExpr> block_size,
                IndexExpr num_nnz_blocks,
                Array<IndexExpr> strides,
                Array<IndexExpr> padding,
                Array<IndexExpr> dilation,
                int groups,
                IndexExpr channels,
                Array<IndexExpr> kernel_size,
                std::string data_layout,
                std::string kernel_layout,
                std::string out_layout,
                DataType out_dtype) {
  auto attrs = make_object<SparseConv2DAttrs>();
  attrs->block_size = std::move(block_size);
  attrs->num_nnz_blocks = std::move(num_nnz_blocks);
  attrs->strides = std::move(strides);
  attrs->padding = std::move(padding);
  attrs->dilation = std::move(dilation);
  attrs->groups = groups;
  attrs->channels = std::move(channels);
  attrs->kernel_size = std::move(kernel_size);
  attrs->data_layout = std::move(data_layout);
  attrs->kernel_layout = std::move(kernel_layout);
  attrs->out_layout = std::move(out_layout);
  attrs->out_dtype = std::move(out_dtype);
  static const Op& op = Op::Get("nn.sparse_conv2d");
  return CallNode::make(op, {data, weight_data, weight_indices, weight_indptr}, Attrs(attrs), {});
}


TVM_REGISTER_GLOBAL("relay.op.nn._make.sparse_conv2d")
.set_body_typed(MakeSparseConv2D);


RELAY_REGISTER_OP("nn.sparse_conv2d")
.describe(R"code(Sparse 2D convolution layer (e.g. spatial convolution over images).

This layer creates a convolution kernel that is convolved
with the layer input to produce a tensor of outputs.

- **data**: This depends on the `layout` parameter. Input is 4D array of shape
            (batch_size, in_channels, height, width) if `layout` is `NCHW`.
- **weight_data**: (num_nnz_blocks, block_size[0], block_size[1] * kernel_size[0] * kernel_size[1])
- **weight_indices**: (num_nnz_blocks) 
- **weight_indptr**: (out_channels // block_size[0] + 1)
- **out**:  This depends on the `layout` parameter. Output is 4D array of shape
            (batch_size, channels, out_height, out_width) if `layout` is `NCHW`.

)code" TVM_ADD_FILELINE)
.set_attrs_type<SparseConv2DAttrs>()
.set_num_inputs(4)
.add_argument("data", "Tensor", "The input tensor.")
.add_argument("weight_data", "1D Tensor", "Weight tensor matrix.")
.add_argument("weight_indices", "1D Tensor", "Weight indices matrix.")
.add_argument("weight_indptr", "1D Tensor", "Weight indptr matrix.")
.set_support_level(2)
.add_type_rel("SparseConv2D", SparseConv2DRel<SparseConv2DAttrs>)
.set_attr<FInferCorrectLayout>("FInferCorrectLayout", SparseConvInferCorrectLayout<SparseConv2DAttrs>);


}  // namespace relay
}  // namespace tvm
